packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/knowledge/KnowledgeController.java

```diff
@@ -106,6 +122,38 @@ ResponseEntity<ExtractionResponse> postCodeToAMR(
 						.getBody());
 	}
 
+		// Create a model from code blocks
+	@PostMapping(value = "/code-blocks-to-model", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
+	@Secured(Roles.USER)
+	public ResponseEntity<ExtractionResponse> codeBlocksToModel(@RequestPart Code code, @RequestPart("file") MultipartFile input) throws IOException {
+
+		try(final CloseableHttpClient httpClient = HttpClients.custom()
+		.build()) {
+		// 1. create code asset from code blocks
+		Code createdCode = codeService.createCode(code);
+
+		// 2. upload file to code asset
+		byte[] fileAsBytes = input.getBytes();
+		HttpEntity fileEntity = new ByteArrayEntity(fileAsBytes, ContentType.APPLICATION_OCTET_STREAM);
+		
+		// we have pre-formatted the files object already so no need to use uploadCode
+		final PresignedURL presignedURL = codeService.getUploadUrl(createdCode.getId(), "tempFile");
+		final HttpPut put = new HttpPut(presignedURL.getUrl());
+		put.setEntity(fileEntity);
+		final HttpResponse response = httpClient.execute(put);
+
+		if(response.getStatusLine().getStatusCode() != 200) {
+			throw new Exception("Error uploading file");
+		}
+		// 3. create model from code asset
+		return ResponseEntity.ok(knowledgeMiddlewareProxy.postCodeToAMR(createdCode.getId().toString(), "temp model", "temp model description", false, false).getBody());
+		} 
+		catch (Exception e) {
+			log.error("unable to upload file", e);
+			return ResponseEntity.internalServerError().build();
```

Comment:
Instead of `return`ing an error, we've moved towards throwing a `ResponseStatusException` as it lets us include an error for the front end:

```
throw new ResponseStatusException(
					org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR,
					"YOUR PLAINTEXT ERROR MESSAGE HERE");
```

Reply:
added 👍